flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java

```diff
@@ -312,9 +313,11 @@ public static Object toFlinkObject(ObjectInspector inspector, Object data) {
 			ListObjectInspector listInspector = (ListObjectInspector) inspector;
 			List<?> list = listInspector.getList(data);
 
-			Object[] result = new Object[list.size()];
+			// flink expects a specific array type (e.g. Integer[] instead of Object[]), so we have to get the element class
+			ObjectInspector elementInspector = listInspector.getListElementObjectInspector();
+			Object[] result = (Object[]) Array.newInstance(classForObjectInspector(elementInspector), list.size());
```

Comment:
with the assumption we discussed in mind, seems that we can just call `Object[] result = (Object[]) Array.newInstance(toFlinkObject(elementInspector, list.get(0).getClass()), list.size());`?

Reply:
Maybe I missed something, but Array.newInstance() needs a Class, while toFlinkObject() returns an Object, right?

Reply:
it's `toFlinkObject(elementInspector, list.get(0)).getClass()`

Reply:
@bowenli86 I'm afraid it won't work because Hive allows empty arrays, in which case we can't get an element instance.